unix,win: add uv_getenv() and uv_setenv()

cjihrig opened the pull request:
These functions are used to create, update, and retrieve environment variables.

Closes #1198

santigimeno commented on src/win/util.c:
```diff
@@ -1387,3 +1390,146 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
+int uv_setenv(const char* name, const char* value) {
+  wchar_t* name_w;
+  wchar_t* value_w;
+  DWORD bufsize;
```
Should `name_w` and `value_w` be initialized to `NULL` so `out` works correctly on every case?

saghul replied:
they are, below, right?

santigimeno replied:
Now, they are. IIRC they were not in the initial commit.

santigimeno commented on src/win/util.c:
```diff
@@ -1387,3 +1390,146 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
+    return uv_translate_sys_error(r);
+  } else if (len > MAX_ENV_VAR_LENGTH) {
+    /* This should not be possible */
+    return UV_EIO;
```
If it shouldn't be possible, would an assertion work here?

cjihrig replied:
An assertion would work. We do this in a few other places though. I don't have a strong preference either way.

santigimeno replied:
I'm fine either way too

saghul replied:
I'd go for an assert. If we ever see it blow up IRL we can replace it with another error.

santigimeno commented on src/unix/core.c:
```diff
@@ -1240,3 +1240,41 @@ int uv_translate_sys_error(int sys_errno) {
+
+  if (var == NULL) {
+    return -ENOENT;
+  }
```
style: remove braces

saghul commented on src/unix/core.c:
```diff
@@ -1240,3 +1240,40 @@ int uv_translate_sys_error(int sys_errno) {
+  }
+
+  memcpy(buffer, var, len + 1);
+  *size = len;
```
please double check if we do this or include the null byte count in other places.

cjihrig replied:
Yea, last year we decided on:

- On `UV_ENOBUFS` set `size` to the required space, including the null.
- On success, set `size` to the string length and null terminate the string.

I don't know if there are functions that still don't follow this, but there were at least a few PRs to migrate the codebase toward this.

saghul replied:
Okidoki!

saghul commented on src/win/util.c:
```diff
@@ -1387,3 +1390,148 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
+    return uv_translate_sys_error(GetLastError());
+
+  /* Convert the environment variable name to a wide character string */
+  name_w = (wchar_t*) uv__malloc(sizeof(wchar_t) * bufsize);
```
unnecessary cast

saghul commented on src/win/util.c:
```diff
@@ -1387,3 +1390,148 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
+  }
+
+  /* Convert the environment variable name to a wide character string */
+  name_w = (wchar_t*) uv__malloc(sizeof(wchar_t) * bufsize);
```
unnecessary cast

saghul commented on src/win/util.c:
```diff
@@ -1387,3 +1390,148 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
+  }
+
+  /* Convert the environment variable value to a wide character string */
+  value_w = (wchar_t*) uv__malloc(sizeof(wchar_t) * bufsize);
```
ditto

saghul commented on src/win/util.c:
```diff
@@ -1387,3 +1390,148 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
+    r = uv_translate_sys_error(GetLastError());
+
+out:
+  if (name_w)
```
no need, free(NULL) is ok

saghul commented on src/win/util.c:
```diff
@@ -1387,3 +1390,148 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
+  if (name_w)
+    uv__free(name_w);
+
+  if (value_w)
```
ditto

saghul commented:
> I'll create a follow up PR for uv_unsetenv() if you're OK with that.

If I may :-) I'd like to have it on this one... that way the feature is complete.

saghul commented:
Naming bikeshed: should ve have these be `uv_os_setenv` and the like not to overpopulate the `uv_` namespace?

santigimeno commented on src/unix/core.c:
```diff
@@ -1240,3 +1240,51 @@ int uv_translate_sys_error(int sys_errno) {
+  if (name == NULL)
+    return -EINVAL;
+
+  if (unsetenv(name) != 0)
```
I wonder if checking for the `name` being `NULL` is really necessary as `unsetenv()` already does that. The same could be said of `setenv()`.

santigimeno commented on src/win/util.c:
```diff
@@ -1387,3 +1390,180 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
+  }
+
+  bufsize = MultiByteToWideChar(CP_UTF8, 0, name, -1, name_w, bufsize);
+
```
Couldn't this `MultiByteToWideChar` / `malloc` / `MultiByteToWideChar` operation be moved to a separate helper function as it is repeated multiple times?

cjihrig replied:
I'm in the process of doing that already :-)

cjihrig commented:
@santigimeno @saghul all existing nits addressed, and added `uv_os_unset()` to this PR. I added a second commit that contains a helper function for the UTF8 to UTF16 conversion on windows.

santigimeno commented on src/win/util.c:
```diff
@@ -1387,3 +1431,113 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
+
+  if (r == 0)
+    r = uv_translate_sys_error(GetLastError());
+
```
`return uv_translate_sys_error(GetLastError());` here directly?

santigimeno commented on src/win/util.c:
```diff
@@ -1387,3 +1431,113 @@ int uv__getpwuid_r(uv_passwd_t* pwd) {
+
+  if (r != 0) {
+    return r;
+  }
```
drop braces

cjihrig commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/222/

cjihrig commented:
Only one test failure (unrelated to this PR). So close to a green CI.

saghul commented on docs/src/misc.rst:
```diff
@@ -386,3 +386,38 @@ API
+
+.. c:function:: int uv_os_unsetenv(const char* name)
+
+    Deletes the environment variable specified by `name` with. If no such
```
s/with//
